README.md
# dhtxx-standalone
Standalone DHT11/DHT22 sensor driver, library and Python module for the Raspberry Pi


ChangeLog
2022-01-23  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* VERSION:

	  - Updated to v0.0.1.

	* ChangeLog:

	  - Change log started.


gentar.sh
#!/bin/sh

# Copyright (C) 2022 Vino Fernando Crescini <vfcrescini@gmail.com>
# SPDX-License-Identifier: GPL-3.0-or-later

# 2022-01-22 gentar.sh

# Generates tarball from the given filelist

TMPDIR="/tmp/.gentar.${$}"

if [ ${#} -lt 2 ]; then
  echo "Usage: gentar.sh <filelist> <tarball>" >&2
  exit 1
fi

path_flist=${1}
path_tball=${2}

mkdir -p ${TMPDIR}

cat ${path_flist} | sed -e 's/^\s*//' -e 's/#.*//g' -e '/^\s*$/d' -e 's/\s*:\s*/:/g' | while IFS=":" read f1 f2 f3 f4 f5 f6; do
  if [ "${f1}" = "F" ] || [ "${f1}" = "D" ]; then
      echo "${f2}" | grep -q '[0-7][0-7][0-7][0-7]'
      if [ ${?} -ne 0 ]; then
        echo "Invalid mode: ${f2}" >&2
        exit 2
      fi
      getent passwd "${f3}" > /dev/null 2>&1
      if [ ${?} -ne 0 ]; then
        echo "Invalid UID: ${f3}" >&2
        exit 3
      fi
      getent group "${f4}" > /dev/null 2>&1
      if [ ${?} -ne 0 ]; then
        echo "Invalid GID: ${f4}" >&2
        exit 4
      fi
  fi
  case ${f1} in
    F)
      if [ `echo "${f5}" | cut -c 1` = "/" ]; then
        if [ ! -f ${f5} ] || [ ! -r ${f5} ]; then
          echo "Not a file or not readable: ${f5}" >&2
          exit 5
        fi
        src=${f5}
      else
        if [ ! -f ${f5} ] || [ ! -r ${f5} ]; then
          echo "Not a file or not readable: ${f5}" >&2
          exit 6
        fi
        src=${f5}
      fi
      install -m ${f2} -o ${f3} -g ${f4} -T ${src} ${TMPDIR}/${f6}
      ;;
    D)
      install -m ${f2} -o ${f3} -g ${f4} -d ${TMPDIR}/${f5}
      ;;
    S)
      ln -fs ${f2} ${TMPDIR}/${f3}
      ;;
    *)
      echo "Unknown type: ${f1}" >&2
      exit 7
      ;;
  esac
done

tar -C ${TMPDIR} -czf ${path_tball} .
rm -fr ${TMPDIR}

exit 0


src/gpio.c
// Copyright (C) 2022 Vino Fernando Crescini <vfcrescini@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

// 2021-09-26 gpio.c

// Very simply GPIO library for RPi
// Uses /proc/device-tree/soc/ranges to get base address; info from:
//   https://github.com/devicetree-org/devicetree-specification/releases/download/v0.3/devicetree-specification-v0.3.pdf


#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <sys/mman.h>

# include "gpio.h"


// get base address from DHTXX_GPIO_DTREE_PATH

static int dhtxx_gpio_base(uint32_t *base, char errstr[DHTXX_GPIO_ERRSTR_MAX]);


// get base address from DHTXX_GPIO_DTREE_PATH

static int dhtxx_gpio_base(uint32_t *base, char errstr[DHTXX_GPIO_ERRSTR_MAX])
{
  int fd = 0;
  int rv = 0;
  uint8_t buf[4];

  memset(buf, 0x00, 4);

  if (base == NULL)
  {
    if (errstr != NULL)
    {
      snprintf(errstr, DHTXX_GPIO_ERRSTR_MAX, "Invalid argument");
    }
    return 1;
  }

  if ((fd = open(DHTXX_GPIO_DTREE_PATH, O_RDONLY)) == -1)
  {
    if (errstr != NULL)
    {
      int e = errno;

      snprintf(errstr, DHTXX_GPIO_ERRSTR_MAX, "Failed to call open(%s): [%d]: %s", DHTXX_GPIO_DTREE_PATH, e, strerror(e));
    }

    return 2;
  }

  // first four bytes are for the child bus address, which we skip over

  if (lseek(fd, 0x04, SEEK_SET) == -1)
  {
    if (errstr != NULL)
    {
      int e = errno;

      snprintf(errstr, DHTXX_GPIO_ERRSTR_MAX, "Failed to call lseek(): [%d]: %s", e, strerror(e));
    }
    return 3;
  }

  // second four bytes are for the parent bus address, which is what we want

  rv = read(fd, &buf, 4);

  if (rv != 4)
  {
    if (rv == -1)
    {
      if (errstr != NULL)
      {
        int e = errno;

        snprintf(errstr, DHTXX_GPIO_ERRSTR_MAX, "Failed to call read(): [%d]: %s", e, strerror(e));
      }
      return 4;
    }
    if (errstr != NULL)
    {
      snprintf(errstr, DHTXX_GPIO_ERRSTR_MAX, "Failed to call read()");
    }
    return 5;
  }

  close(fd);

  // because some machines are little endian

  *base = buf[0] << 24 | buf[1] << 16 | buf[2] << 8 | buf[3] << 0;

  return 0;

}


// create a memory map of the GPIOs, and give back a pointer to it
// returns zero on success, non-zero otherwise, with errstr set

int dhtxx_gpio_init(volatile uint32_t **gpio, char errstr[DHTXX_GPIO_ERRSTR_MAX])
{
  int fd;
  uint32_t base;
  volatile uint32_t *tmp = NULL;

  if (gpio == NULL)
  {
    if (errstr != NULL)
    {
      snprintf(errstr, DHTXX_GPIO_ERRSTR_MAX, "Invalid argument");
    }
    return 1;
  }

  if (dhtxx_gpio_base(&base, errstr))
  {
    return 2;
  }

  fd = open("/dev/mem", O_RDWR | O_SYNC);

  if (fd < 0)
  {
    if (errstr != NULL)
    {
      int e = errno;
      snprintf(errstr, DHTXX_GPIO_ERRSTR_MAX, "Failed call to open(/dev/mem): [%d] %s", e, strerror(e));
    }
    return 3;
  }

  tmp = (uint32_t *) mmap(NULL, DHTXX_GPIO_LEN, PROT_READ | PROT_WRITE, MAP_SHARED, fd, base + DHTXX_GPIO_BASE_OFFSET);

  if (tmp == MAP_FAILED)
  {
    if (errstr != NULL)
    {
      int e = errno;
      snprintf(errstr, DHTXX_GPIO_ERRSTR_MAX, "Failed call to mmap(/dev/mem): [%d] %s", e, strerror(e));
    }

    close(fd);

    return 4;
  }

  close(fd);

  *gpio = tmp;

  return 0;
}


// set GPIO pin for input

int dhtxx_gpio_set_input(volatile uint32_t *gpio, int num)
{
  volatile uint32_t *tmp;

  if (gpio == NULL || num < 0)
  {
    return 1;
  }

  tmp = gpio + (num / 10);
  *tmp = *tmp & ~(7 << ((num % 10) * 3));

  return 0;
}


// set GPIO pin for output

int dhtxx_gpio_set_output(volatile uint32_t *gpio, int num)
{
  volatile uint32_t *tmp;

  if (gpio == NULL || num < 0)
  {
    return 1;
  }

  tmp = gpio + (num / 10);
  *tmp = *tmp & ~(7 << ((num % 10) * 3));
  *tmp = *tmp | (1 << ((num % 10) * 3));

  return 0;
}


// sets value to the current state of the given input GPIO pin
// value is set to non-zero if pin is high, zero otherwise
// returns zero on success, non-zero otherwise

int dhtxx_gpio_read(volatile uint32_t *gpio, int num, uint8_t *value)
{
  if (gpio == NULL || num < 0 || value == NULL)
  {
    return 1;
  }

  *value = (*(gpio + 13) & (1 << num)) != 0;

  return 0;
}


// asserts/deasserts given output GPIO pin
// assert if value != 0, deassert if value == 0
// returns zero on success, non-zero otherwise

int dhtxx_gpio_write(volatile uint32_t *gpio, int num, uint8_t value)
{
  if (gpio == NULL || num < 0)
  {
    return 1;
  }

  *(gpio + (value ? 7 : 10)) = 1 << num;

  return 0;
}


src/gpio.h
// Copyright (C) 2022 Vino Fernando Crescini <vfcrescini@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

// 2021-09-26 gpio.h

// Very simply GPIO library for RPi
// Uses /proc/device-tree/soc/ranges to get base address; info from:
//   https://github.com/devicetree-org/devicetree-specification/releases/download/v0.3/devicetree-specification-v0.3.pdf


# include <stdint.h>


#define DHTXX_GPIO_DTREE_PATH  "/proc/device-tree/soc/ranges"
#define DHTXX_GPIO_BASE_OFFSET 0x200000
#define DHTXX_GPIO_LEN         4096
#define DHTXX_GPIO_ERRSTR_MAX  128


// create a memory map of the GPIOs, and give back a pointer to it
// returns zero on success, non-zero otherwise, with errstr set

int dhtxx_gpio_init(volatile uint32_t **gpio, char errstr[DHTXX_GPIO_ERRSTR_MAX]);


// set GPIO pin for input

int dhtxx_gpio_set_input(volatile uint32_t *gpio, int num);


// set GPIO pin for output

int dhtxx_gpio_set_output(volatile uint32_t *gpio, int num);


// sets value to the current state of the given input GPIO pin
// value is set to non-zero if pin is high, zero otherwise
// returns zero on success, non-zero otherwise

int dhtxx_gpio_read(volatile uint32_t *gpio, int num, uint8_t *value);


// asserts/deasserts given output GPIO pin
// assert, if value != 0, deassert if value == 0
// returns zero on success, non-zero otherwise

int dhtxx_gpio_write(volatile uint32_t *gpio, int num, uint8_t value);


src/main.c
// Copyright (C) 2022 Vino Fernando Crescini <vfcrescini@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

// 2021-09-26 main.c

// Standalone DHT11/DHT22 sensor driver for Raspberry PI


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>

#include "gpio.h"
#include "dhtxx.h"


int main(int argc, char *argv[])
{
  int pin = 0;
  int curr_opt = 0;
  char stype = DHTXX_TYPE_DHT22;
  float ohum = 0.0;
  float otmp = 0.0;
  char errstr[DHTXX_ERRSTR_MAX];

  opterr = 0;
  optind = 0;

  while((curr_opt = getopt(argc, argv, "hd")) != -1)
  {
    switch(curr_opt)
    {
      case 'h':
        fprintf(stderr, "Usage: dhtxx [-h|-d] <pin>\n");
        return 0;
        break;
      case 'd':
        stype = DHTXX_TYPE_DHT11;
        break;
      default:
        fprintf(stderr, "Invalid option: -%c\n", curr_opt);
        return 1;
    }
  }

  if (argc <= optind)
  {
    fprintf(stderr, "Need to specify GPIO pin\n");
    return 2;
  }

  pin = atoi(argv[optind]);

  if (pin <= 0 || pin > 40)
  {
    fprintf(stderr, "Invalid GPIO pin: %d\n", pin);
    return 3;
  }

  memset(errstr, 0x00, DHTXX_ERRSTR_MAX);

  if (dhtxx_get(pin, stype, &ohum, &otmp, errstr))
  {
    fprintf(stderr, "Failed to get data: %s\n", errstr);
    return 4;
  }

  // dump result

  fprintf(stdout, "%8.3f %8.3f\n", ohum, otmp);

  return 0;
}


src/dhtxx.c
// Copyright (C) 2022 Vino Fernando Crescini <vfcrescini@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

// 2021-09-26 dhtxx.c

// Standalone DHT11/DHT22 sensor driver for Raspberry PI
// Data from:
//  https://cdn.datasheetspdf.com/pdf-down/D/H/T/DHT11-Aosong.pdf
//  https://cdn.datasheetspdf.com/pdf-down/D/H/T/DHT22-Aosong.pdf


#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <sched.h>
#include <errno.h>
#include <time.h>
#include <sys/time.h>

#include "gpio.h"
#include "dhtxx.h"


#define DHTXX_TICK_MAX 32000


static int dhtxx_set_priority_max();

static int dhtxx_set_priority_normal();

static int dhtxx_millisleep(uint32_t msec);

static int dhtxx_milliwait(uint32_t msec);


static int dhtxx_set_priority_max()
{
  struct sched_param sp;

  memset(&sp, 0x00, sizeof(sp));

  sp.sched_priority = sched_get_priority_max(SCHED_FIFO);

  if (sched_setscheduler(0, SCHED_FIFO, &sp) == -1)
  {
    return 1;
  }

  return 0;
}


static int dhtxx_set_priority_normal()
{
  struct sched_param sp;

  memset(&sp, 0x00, sizeof(sp));

  sp.sched_priority = 0;

  if (sched_setscheduler(0, SCHED_OTHER, &sp) == -1)
  {
    return 1;
  }

  return 0;
}


static int dhtxx_millisleep(uint32_t msec)
{
  struct timespec st;

  memset(&st, 0x00, sizeof(st));

  // msec = 1 / 1000 secs
  // nsec  = 1 / 1000000000 secs = 1 / 1000000 usec

  st.tv_sec = msec / 1000;
  st.tv_nsec = (msec % 1000) * 1000000;

  while(1)
  {
    if (clock_nanosleep(CLOCK_MONOTONIC, 0, &st, &st) != 0 && errno != EINTR)
    {
      return 1;
    }

    break;
  }

  return 0;
}


static int dhtxx_milliwait(uint32_t msec)
{
  struct timeval stc;
  struct timeval std;
  struct timeval ste;

  memset(&stc, 0x00, sizeof(stc));
  memset(&std, 0x00, sizeof(std));
  memset(&ste, 0x00, sizeof(ste));

  // msec = 1 / 1000 secs
  // usec  = 1 / 1000000 secs = 1 / 1000 msec

  std.tv_sec = msec / 1000;
  std.tv_usec = (msec % 1000) * 1000;

  gettimeofday(&stc, NULL);

  timeradd(&stc, &std, &ste);

  while(1)
  {
    gettimeofday(&stc, NULL);

    if (timercmp(&stc, &ste, >))
    {
      break;
    }
  }

  return 0;
}


// get humidity and temperature reading from sensor of the given type

int dhtxx_get(int pin, dhtxx_type_t stype, float *ahum, float *atmp, char errstr[DHTXX_ERRSTR_MAX])
{ 
  volatile uint32_t *gpio = NULL;
  uint32_t tick_average_low = 0;
  uint32_t tick_count_low[41];
  uint32_t tick_count_high[41];
  uint8_t data[5];
  float ohum = 0.0;
  float otmp = 0.0;

  memset(tick_count_low, 0x00, sizeof(tick_count_low));
  memset(tick_count_high, 0x00, sizeof(tick_count_high));
  memset(data, 0x00, sizeof(data));

  if (pin <= 0 || pin > 40 || (stype != DHTXX_TYPE_DHT11 && stype != DHTXX_TYPE_DHT22) || ahum == NULL || atmp == NULL)
  {
    if (errstr != NULL)
    {
      snprintf(errstr, DHTXX_ERRSTR_MAX, "Invalid arguments");
    }
    return 1;
  }

  if (dhtxx_gpio_init(&gpio, errstr))
  {
    return 2;
  }

  dhtxx_set_priority_max();

  // step 1: set pin to output, assert for 500 msec, then de-assert for 20 msec

  dhtxx_gpio_set_output(gpio, pin);

  dhtxx_gpio_write(gpio, pin, 1);

  dhtxx_millisleep(500);

  dhtxx_gpio_write(gpio, pin, 0);

  dhtxx_milliwait(20);

  // step 2: set pin input, then sleep

  dhtxx_gpio_set_input(gpio, pin);

  // XXX - there's gotta be a better way
  // declare as volatile so it doesn't get optimised out despite being no-op

  for(volatile uint8_t i = 0; i < 50; i++);

  // step 3: wait for sensor to de-assert pin

  for(uint32_t i = 0; 1; i++)
  {
    uint8_t value = 0;

    if (i >= DHTXX_TICK_MAX)
    {
      if (errstr != NULL)
      {
        snprintf(errstr, DHTXX_ERRSTR_MAX, "Timeout while waiting for sensor to de-assert pin at initalisation");
      }
      return 3;
    }

    dhtxx_gpio_read(gpio, pin, &value);

    if (value == 0)
    {
      break;
    }
  }

  // step 4: record pulses

  for(uint8_t i = 0; i < 41; i++)
  {
    uint8_t value = 0;

    // count ticks until sensor asserts pin

    while(1)
    {
      dhtxx_gpio_read(gpio, pin, &value);

      if (value != 0)
      {
        break;
      }

      tick_count_low[i] = tick_count_low[i] + 1;

      if (tick_count_low[i] >= DHTXX_TICK_MAX)
      {
	if (errstr != NULL)
        {
          snprintf(errstr, DHTXX_ERRSTR_MAX, "Timeout while waiting for sensor to assert pin at bit %d", i);
	}
        return 4;
      }
    }

   // count ticks until sensor de-asserts pin

    while(1)
    {
      dhtxx_gpio_read(gpio, pin, &value);

      if (value == 0)
      {
        break;
      }

      tick_count_high[i] = tick_count_high[i] + 1;

      if (tick_count_high[i] >= DHTXX_TICK_MAX)
      {
	if (errstr != NULL)
        {
          snprintf(errstr, DHTXX_ERRSTR_MAX, "Timeout while waiting for sensor to de-assert pin at bit %d", i);
	}
        return 5;
      }
    }
  }

  // done. now drop back to normal priority

  dhtxx_set_priority_normal();

  // according to specs, low pulses are 50 msecs long
  // high pulses that are short (26 to 26 msecs) represent 0
  // high pulses that are long (70 msecs) represent 1

  // calculate average low signal ticks

  for(uint8_t i = 0; i < 40; i++)
  {
    tick_average_low = tick_average_low + tick_count_low[i + 1];
  }

  tick_average_low = tick_average_low / 40;

  // use low average tick count to decide whether high tick count is a 0 or a 1
  // skip first bit
  // remaining 40 bits = 5 bytes

  for(uint8_t i = 0; i < 5; i++)
  {
    for(uint8_t j = 0; j < 8; j++)
    {
      uint8_t bit = tick_count_high[(i * 8) + j + 1] >= tick_average_low;

      data[i] = (data[i] << 1) | bit;
    }
  }

  // validate checksum
  // fifth byte should be equal to the lowest eight bits of the sum of the first four bytes

  if (data[4] != ((data[0] + data[1] + data[2] + data[3]) & 0xff))
  {
    if (errstr != NULL)
    {
      snprintf(errstr, DHTXX_ERRSTR_MAX, "Checksum error");
    }
    return 6;
  }

  // calculate

  switch(stype)
  {
    case DHTXX_TYPE_DHT11:
      ohum = data[0] + (data[1] * 0.1);
      otmp = data[2] + (data[3] * 0.1);
      break;
    case DHTXX_TYPE_DHT22:
    default:
      ohum = ((data[0] << 8) + data[1]) / 10.0;
      otmp = (((data[2] & 0x7F) << 8) + data[3]) * ((data[2] & 0x80) ? -0.1 : 0.1);
      break;
  }

  // validate range

  switch(stype)
  {
    case DHTXX_TYPE_DHT11:
      if (ohum < DHTXX_DHT11_HUM_MIN || ohum > DHTXX_DHT11_HUM_MAX)
      {
	if (errstr != NULL)
        {
          snprintf(errstr, DHTXX_ERRSTR_MAX, "Humidity out of range: %8.3f", ohum);
	}
        return 7;
      }

      if (otmp < DHTXX_DHT11_TMP_MIN || otmp > DHTXX_DHT11_TMP_MAX)
      {
       	if (errstr != NULL)
        {
          snprintf(errstr, DHTXX_ERRSTR_MAX, "Temperature out of range: %8.3f", otmp);
	}
        return 8;
      }

      break;
    case DHTXX_TYPE_DHT22:
    default:
      if (ohum < DHTXX_DHT22_HUM_MIN || ohum > DHTXX_DHT22_HUM_MAX)
      {
	if (errstr != NULL)
        {
          snprintf(errstr, DHTXX_ERRSTR_MAX, "Humidity out of range: %8.3f", ohum);
	}
        return 7;
      }

      if (otmp < DHTXX_DHT22_TMP_MIN || otmp > DHTXX_DHT22_TMP_MAX)
      {
       	if (errstr != NULL)
        {
          snprintf(errstr, DHTXX_ERRSTR_MAX, "Temperature out of range: %8.3f", otmp);
	}
        return 8;
      }

      break;
  }

  *ahum = ohum;
  *atmp = otmp;

  return 0;
}


src/dhtxx.h
// Copyright (C) 2022 Vino Fernando Crescini <vfcrescini@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

// 2021-09-26 dhtxx.h

// Standalone DHT11/DHT22 sensor driver for Raspberry PI
// Data from:
//  https://cdn.datasheetspdf.com/pdf-down/D/H/T/DHT11-Aosong.pdf
//  https://cdn.datasheetspdf.com/pdf-down/D/H/T/DHT22-Aosong.pdf


#define DHTXX_DHT11_HUM_MIN  20.000
#define DHTXX_DHT11_HUM_MAX  90.000
#define DHTXX_DHT11_TMP_MIN   0.000
#define DHTXX_DHT11_TMP_MAX  50.000

#define DHTXX_DHT22_HUM_MIN   0.000
#define DHTXX_DHT22_HUM_MAX 100.000
#define DHTXX_DHT22_TMP_MIN -40.000
#define DHTXX_DHT22_TMP_MAX  80.000

#define DHTXX_ERRSTR_MAX    128


typedef enum
{
  DHTXX_TYPE_DHT11,
  DHTXX_TYPE_DHT22
}  dhtxx_type_t;


// get humidity and temperature reading from sensor of the given type

int dhtxx_get(int pin, dhtxx_type_t stype, float *ahum, float *atmp, char errstr[DHTXX_ERRSTR_MAX]);


src/modpython.c
// Copyright (C) 2022 Vino Fernando Crescini <vfcrescini@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

// 2022-01-19 modpython.c

// python module to expose libdhtxx functions


#include <stdlib.h>
#include <string.h>
#include <Python.h>

#include "dhtxx.h"


static PyObject *_dhtxx_get(PyObject* self, PyObject* args);


static PyMethodDef dhtxx_methods[] =
{
  { "get", _dhtxx_get, METH_VARARGS, "Get humidity and temperature data from DHTXX sensor a the given GPIO pin" },
  { NULL, NULL, 0, NULL }
};

static struct PyModuleDef dhtxx_module =
{
  PyModuleDef_HEAD_INIT,
  "dhtxx",
  "Simple DHTXX modle",
  -1,
  dhtxx_methods
};


PyMODINIT_FUNC PyInit_dhtxx(void)
{

  return PyModule_Create(&dhtxx_module);
}


static PyObject * _dhtxx_get(PyObject* self, PyObject* args)
{
  int pin = 0;
  int stype = 0;
  float ohum = 0.0;
  float otmp = 0.0;
  char oerrstr[DHTXX_ERRSTR_MAX];

  if (!PyArg_ParseTuple(args, "ii", &pin, &stype))
  {
    PyErr_SetString(PyExc_RuntimeError, "Invalid arguments");
    return NULL;
  }

  memset(oerrstr, 0x00, DHTXX_ERRSTR_MAX);

  if (dhtxx_get(pin, stype, &ohum, &otmp, oerrstr) != 0)
  {
    PyErr_Format(PyExc_RuntimeError, "Get failed: %s", oerrstr);
    return NULL;
  }

  return Py_BuildValue("(f,f)", ohum, otmp);
}
